### ISSUE TITLE
The huge problem is that I need to UPDATE the cookies so that I have a valid access token at the next request

### ISSUE THREAD

**Issue by sanskar-19** [vercel/next.js#54266]
The huge problem is that I need to UPDATE the cookies so that I have a valid access token at the next request

_Originally posted by @ndt080 in https://github.com/vercel/next.js/issues/47126#issuecomment-1469480240_

**Comment by sanskar-19** [comment]
So I want to fetch data on the server with a token stored on the client browser side. 

Now cookies are being shared from browser -> server -> route handler. For if my token expires I want to refresh the same and update the cookie on the browser too with the latest token.

The problem I face is the nextjs Page.tsx will ignore the set-Cookie header. as it is meant for the browser. So how do I update the cookie for the same.

**Comment by masterbater** [comment]
> So I want to fetch data on the server with a token stored on the client browser side.
> 
> Now cookies are being shared from browser -> server -> route handler. For if my token expires I want to refresh the same and update the cookie on the browser too with the latest token.
> 
> The problem I face is the nextjs Page.tsx will ignore the set-Cookie header. as it is meant for the browser. So how do I update the cookie for the same.

middleware

**Comment by sanskar-19** [comment]
> > So I want to fetch data on the server with a token stored on the client browser side.
> > Now cookies are being shared from browser -> server -> route handler. For if my token expires I want to refresh the same and update the cookie on the browser too with the latest token.
> > The problem I face is the nextjs Page.tsx will ignore the set-Cookie header. as it is meant for the browser. So how do I update the cookie for the same.
> 
> middleware

Umm, so middleware would be able to intercept it. The route has handled stuff on the server itself. How can I send that set-cookie header back to the browser.

I don't think middleware would be able to parse it out like that. And how do I call the middleware when called the route handler in the server component

**Comment by ndt080** [comment]
@sanskar-19 
Most of the responses with a link to middleware are invalid, since people think that the authorization api is part of your next.js applications To date, there is no complete solution, but you can make a "crutch"

The main problem is updating the token on the server side, so you will have to abandon the standard jwt token refresh mechanism.

I wrote an abstraction that I use in middleware every time we send a request from the browser to the server to receive an HTML page. At this point, I'm parsing the JWT and checking the token lifetime, if the token has expired, I update it and set a header in the Set Cookie, which will return to the client and update its local cookies.

My **middlewate.ts**
https://github.com/darkguin/darktheater-next/blob/main/src/middleware.ts
```
const middlewares: Middleware[] = [
  AuthMiddleware,
  UnAuthMiddleware,
  CatalogMiddleware,
  CollectionsMiddleware,
  SessionMiddleware,
];

export async function middleware(request: NextRequest) {
  const response = NextResponse.next();

  const isPage = RoutesList.includes(request.nextUrl.pathname as Route);
  const nextResponse = isPage ? await withMiddleware(request, response, middlewares) : null;

  return nextResponse || response;
}

export const config = {
  matches: ['/**'],
};
```
My **session-middleware.ts**
https://github.com/darkguin/darktheater-next/blob/main/src/processes/auth/middlewares/session-middleware.ts
```
export const SessionMiddleware = defineMiddleware({
  global: true,
  async handler(req: NextRequest) {
    const accessToken = req.cookies.get(StorageKey.AccessToken)?.value;
    const refreshToken = req.cookies.get(StorageKey.RefreshToken)?.value;

    if (!accessToken || !refreshToken) return;

    const { exp } = decodeJwt<TokenPayload>(accessToken) ?? {};
    const isValid = dayjs().isBefore(unix(exp!), 'minute');

    if (isValid) return;

    try {
      console.log('INFO: Start refresh tokens');
      const { refreshSession } = sessionApi();
      const session = await refreshSession(refreshToken);
      console.log('INFO: End refresh tokens', session);

      const response = NextResponse.redirect(new URL(req.url));
      response.cookies.set(StorageKey.AccessToken, session.accessToken);
      response.cookies.set(StorageKey.RefreshToken, session.refreshToken);
      return response;
    } catch (e) {
      console.log('Error: Error during token refresh', e);
      const response = NextResponse.redirect(new URL(Route.SignIn, req.url));
      response.cookies.delete(StorageKey.AccessToken);
      response.cookies.delete(StorageKey.RefreshToken);
      return response;
    }
  },
});
```

**Comment by ndt080** [comment]
It is worth noting that on the client side, the JWT token is updated in accordance with the standard mechanism

**Comment by sanskar-19** [comment]
But this is something that you are deliberately. You are consistently refreshing your token. I mean with the pages router we could have passed the token as props. But with the app directory since the page component is also server side and route handler too how can you pass that token piece to the client. Had this been a client component it is getting updated on the browser. 

We are able to refresh the token, but the token rotation cycle is not possible. We cannot sync the token up considering we send a token which is expired to the server component.

**Comment by ndt080** [comment]
The middleware runs before accessing the page component. That's why the only option for this component to have an up-to-date token when sending API requests from the server is to update it in the middleware. Thus, we keep the token rotation cycle

**Comment by sanskar-19** [comment]
So basically, you are suggesting to constantly refresh out the token whilst accessing protected component to make sure that the token is expired neither on the client nor on the server. 

But for incase if the token expiry is maybe too small then in that particular case, lets consider a wait or a delay maybe, so I guess then you have to use a custom axios instance to refresh it on the server and on the client calls too. 

(Considering you have a very very less token expiry).

**Comment by ndt080** [comment]
you don't need to constantly update the token in middleware. It will be updated only if its token lifetime has expired. As a rule, in 80% of the situation, the token will be updated on the client side, and in exceptional cases - on the server side (for example, the initial opening of the application), and the updated token from the server will be returned to the client.


Unfortunately, the idea with an axios instance on the server side will not succeed, since requests through it will be sent from the React Server Component, which means that you will not be able to return an already updated token to the client

You cannot use the structure that you stored separate instances of tokens on the server side, since there is no common context between pages and server entities.

**Comment by sanskar-19** [comment]
I have a backend auth service separately, so you want to basically de-structure the payload and check the expiry. hmm. but maybe I don't want to do that. 

See the problem is in the middleware I am not expecting it to constantly keep a check onto my token for its validity since I am relying on a python based ASGI server for the same. So I cannot just de-structure the same. 

I want it to get refreshed on the time when it expires. I agree that most calls be client side for interactivity but let's say there is a view order details section now that is ready only. 

And now like I said if I don't check the validity in the middleware and let route handlers do the job then the token won't be synced on both the client and server. 

Perhaps we can use two different tokens generated from same refresh token just the fact that they both wont be synced.

**Comment by ndt080** [comment]
I only suggested the only solution I found. Unfortunately, your idea with shared tokens will not succeed if you do not create a database from the server side that will store the token. You can't just save the token to a global variable or data structure, since there is no (sometimes intentionally overwritten) global context between calls to the server components of the pages.

**Comment by masterbater** [comment]
> I have a backend auth service separately, so you want to basically de-structure the payload and check the expiry. hmm. but maybe I don't want to do that. 
> 
> See the problem is in the middleware I am not expecting it to constantly keep a check onto my token for its validity since I am relying on a python based ASGI server for the same. So I cannot just de-structure the same. 
> 
> I want it to get refreshed on the time when it expires. I agree that most calls be client side for interactivity but let's say there is a "view order details" section now that is read-only. 
> 
> And now like I said if I don't check the validity in the middleware and let route handlers do the job then the token won't be synced on both the client and server. (Considering I make the call from the server itself). 
> 
> Perhaps we can use two different tokens generated from same refresh token just the fact that they both won't be synced.
> 
> The browser and the backend servers will share different tokens different expiry perhaps for incase. Very rare ones.

Refresh token is the key validate it in client and also set the cookies also in the client or use jwt

**Comment by sanskar-19** [comment]
> I only suggested the only solution I found. Unfortunately, your idea with separeted tokens will not succeed if you do not create a database from the server side that will store the token. You can't just save the token to a global variable or data structure, since there is no (sometimes intentionally overwritten) global context between calls to the server components of the pages.

Agreed that we don't have a global context to share the token, but do we have a need to store it?. I mean we can refresh the same with the refresh token, and so instead of storing it would generate a new token on the server use it and thats it. (for incase if the token sent via cookie is expired.)

**Comment by sanskar-19** [comment]
> > I have a backend auth service separately, so you want to basically de-structure the payload and check the expiry. hmm. but maybe I don't want to do that.
> > See the problem is in the middleware I am not expecting it to constantly keep a check onto my token for its validity since I am relying on a python based ASGI server for the same. So I cannot just de-structure the same.
> > I want it to get refreshed on the time when it expires. I agree that most calls be client side for interactivity but let's say there is a "view order details" section now that is read-only.
> > And now like I said if I don't check the validity in the middleware and let route handlers do the job then the token won't be synced on both the client and server. (Considering I make the call from the server itself).
> > Perhaps we can use two different tokens generated from same refresh token just the fact that they both won't be synced.
> > The browser and the backend servers will share different tokens different expiry perhaps for incase. Very rare ones.
> 
> Refresh token is the key validate it in client and also set the cookies also in the client or use jwt

Client side everything is achievable but with the new concept of nextjs 13, now I am rendering data on the server itself with the token. But token rotation is becoming a problem.

Otherwise client side token rotation is pretty straightforward.

**Comment by ndt080** [comment]
If you do not save the token, then there will be some problems
1. The token received from the server becomes disposable. We will have to receive it several times with several requests to the api from the server. At the same time, we will have to update it again, but on the client side.
2. If you make a special non-expired token for the server, then the security level decreases.
